flocker/control/_model.py

```diff
@@ -983,7 +1017,16 @@ def __new__(cls, **kwargs):
 
     uuid = field(type=UUID, mandatory=True)
     hostname = field(type=unicode, factory=unicode, mandatory=True)
-    applications = pset_field(Application, optional=True, initial=None)
+    applications = pmap_field(
+        unicode, Application, optional=True, initial=None,
+        invariant=_keys_match("name"),
+        factory=(
+            lambda x: (
+                x if x is None
+                else _turn_iterable_to_mapping_from_attribute('name', x)
```

Comment:
Sure, this is needed because here the field is optional, whereas in the other location it is not optional.  But... that is more clearly communicated as a parameter to the factory than a `x if x is None` here.